elements/lisk-p2p/test/integration/p2p.ts

```diff
@@ -1475,4 +1475,82 @@ describe('Integration tests for P2P library', () => {
 			});
 		});
 	});
+
+	describe('Network with peer inbound eviction protection enabled', () => {
+		const NETWORK_PEER_COUNT_WITH_LIMIT = 10;
+		const MAX_INBOUND_CONNECTIONS = 5;
+		const POPULATOR_INTERVAL_WITH_LIMIT = 10;
+		beforeEach(async () => {
+			p2pNodeList = [...new Array(NETWORK_PEER_COUNT_WITH_LIMIT).keys()].map(
+				index => {
+					// Each node will have the previous node in the sequence as a seed peer except the first node.
+					const seedPeers = [
+						{
+							ipAddress: '127.0.0.1',
+							wsPort:
+								NETWORK_START_PORT +
+								((index + 1) % NETWORK_PEER_COUNT_WITH_LIMIT),
+						},
+					];
+
+					const nodePort = NETWORK_START_PORT + index;
+					return new P2P({
+						connectTimeout: 300,
+						ackTimeout: 5000,
+						seedPeers,
+						wsEngine: 'ws',
+						populatorInterval: POPULATOR_INTERVAL_WITH_LIMIT,
+						maxOutboundConnections: MAX_INBOUND_CONNECTIONS,
+						maxInboundConnections: 5,
+						evictionProtectionEnabled: true,
+						nodeInfo: {
+							wsPort: nodePort,
+							nethash:
+								'da3ed6a45429278bac2666961289ca17ad86595d33b31037615d4b8e8f158bba',
+							version: '1.0.1',
+							protocolVersion: '1.0.1',
+							minVersion: '1.0.0',
+							os: platform(),
+							height: 0,
+							broadhash:
+								'2768b267ae621a9ed3b3034e2e8a1bed40895c621bbb1bbd613d92b9d24e54b5',
+							nonce: `O2wTkjqplHII${nodePort}`,
+						},
+					});
+				},
+			);
+
+			// Start nodes incrementally to make inbound eviction behavior predictable
+			p2pNodeList.forEach(async p2p => {
+				await wait(600);
+				p2p.start();
+			});
+			await wait(1000);
+		});
+
+		afterEach(async () => {
+			await Promise.all(
+				p2pNodeList
+					.filter(p2p => p2p.isActive)
+					.map(async p2p => await p2p.stop()),
+			);
+			await wait(100);
+		});
+
+		describe('Inbound peer evictions', () => {
+			it('should evict inbound peer which connected the latest', async () => {
+				p2pNodeList.forEach(async (p2p, i) => {
+					// Disregard early peers who may be re-discovered by already evicted peers
+					if (i <= 3) {
+						return;
+					}
+					const inboundPeers = p2p['_peerPool']
+						.getPeers(InboundPeer)
+						.map(peer => peer.wsPort);
+					const lastPeerPort = 5009;
+					expect(inboundPeers).to.not.include(lastPeerPort);
```

Comment:
I don't get the scenario. We should think of something more robust than this.

Reply:
This scenario is testing specifically for the `connectTime` peer protection category, where peers who connected the earliest are less likely to be evicted.  I am testing for this as it is difficult to simulate peers with varying `latency` as well as `responseRate` in these integration tests, without potentially using something like `rewire`. I agree that we should have more robust tests eventually. Thoughts? @jondubois

Reply:
I've received this error when executing the test:
```
(node:3368) UnhandledPromiseRejectionWarning: AssertionError: expected [ 5003, 5002, 5000, 5001, 5009 ] to not include 5009
    at p2pNodeList.forEach (/Users/diego-g/git/lisk-sdk/elements/lisk-p2p/test/integration/p2p.ts:1551:34)
    at Array.forEach (<anonymous>)
    at Context.it (/Users/diego-g/git/lisk-sdk/elements/lisk-p2p/test/integration/p2p.ts:1542:17)
    at callFn (/Users/diego-g/git/lisk-sdk/elements/lisk-p2p/node_modules/mocha/lib/runnable.js:372:21)
    at Test.Runnable.run (/Users/diego-g/git/lisk-sdk/elements/lisk-p2p/node_modules/mocha/lib/runnable.js:364:7)
    at Runner.runTest (/Users/diego-g/git/lisk-sdk/elements/lisk-p2p/node_modules/mocha/lib/runner.js:455:10)
    at /Users/diego-g/git/lisk-sdk/elements/lisk-p2p/node_modules/mocha/lib/runner.js:573:12
    at next (/Users/diego-g/git/lisk-sdk/elements/lisk-p2p/node_modules/mocha/lib/runner.js:369:14)
    at /Users/diego-g/git/lisk-sdk/elements/lisk-p2p/node_modules/mocha/lib/runner.js:379:7
    at next (/Users/diego-g/git/lisk-sdk/elements/lisk-p2p/node_modules/mocha/lib/runner.js:303:14)
    at Immediate.<anonymous> (/Users/diego-g/git/lisk-sdk/elements/lisk-p2p/node_modules/mocha/lib/runner.js:347:5)
    at runCallback (timers.js:705:18)
    at tryOnImmediate (timers.js:676:5)
    at processImmediate (timers.js:658:5)
(node:3368) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
(node:3368) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

Reply:
This is difficult to test. One way to make it easier/more predictable would be to add support for additional config options to allow us to enable and disable each of the 3 characteristics individually then each test could test just one characteristic and turn off the others:

- Peer latency: We protect a number of peers with optimal latency. An attacker cannot manipulate this metric without physically moving nodes closer to the target.

  - Useful work: We protect peers who perform useful work, such as peers who have recently sent blocks and transactions. An attacker cannot manipulate this metric without performing useful work.

  - Connection time: We protect half of the remaining nodes who have been connected the longest. This precludes attacks that start later.

Reply:
I have created new configurations to enable.disable protection by category (latency, connecttime, response, etc). For the integration test I only enable connectTime, and limit the scope of peers to make the behavior predictable.